src/Service/OpenamApiClient.php

```diff
@@ -169,8 +201,13 @@ public function queryEndpoint(array $options = []) {
       $responseContents = Json::decode($responseContents);
     }
     catch (\Exception $e) {
-      // TODO: Better handling of exception with response status codes.
+      if (isset(static::HTTP_STATUS_CODE_MESSAGES[$e->getCode()])) {
+        $this->logError(static::HTTP_STATUS_CODE_MESSAGES[$e->getCode()], $e);
+        throw $e;
+      }
+
       $this->logError('Error querying the endpoint', $e);
+      throw $e;
```

Comment:
Is it better to throw the error or simply handle it (log)? Or both?

Reply:
Good question i pondered over this for a while and settled on this for the following reasons.

We throw so that it can be caught and acted upon by the method which is calling queryEndpoint().

We log as here we are just logging the error messages that OpenAM give generally. However they will also have different meanings per endpoint and we may want to log which endpoint’s query failed.

For example with the authenticate endpoint if you give incorrect details a 401 is thrown, however a 401 is also thrown if the account is locked.

In this scenario the queryEndpoint() will log the standard 401 error and throw the exception. This exception is caught by the authenticate method. The authenticate method inspects the error message returned. It logs whether it was due to the account being locked or invalid user details. It then throws its own new exception, either 'locked' or 'invalid'. We do this as the request return structure returned by openam for these errors changes between versions. For example In openam12 the messages were different to AM5.5. IMO the module implementing openam_api should not have to be updated when error messages change so we throw a standardised error. Any module implementing openam_api can then catch these standardised exceptions to action any resulting business logic. For example a message to the user saying ‘you account has been locked’ or these 'login details are invalid'.